packages/plugin-typescript/src/lib/runner/runner.ts

```diff
@@ -0,0 +1,60 @@
+import type {
+  AuditOutput,
+  AuditOutputs,
+  AuditReport,
+  Issue,
+  RunnerFunction,
+} from '@code-pushup/models';
+import type { AuditSlug } from '../types.js';
+import {
+  type DiagnosticsOptions,
+  getTypeScriptDiagnostics,
+} from './ts-runner.js';
+import type { CodeRangeName } from './types.js';
+import { getIssueFromDiagnostic, tSCodeToAuditSlug } from './utils.js';
+
+export type RunnerOptions = DiagnosticsOptions & {
+  expectedAudits: { slug: AuditSlug }[];
+};
+
+export function createRunnerFunction(options: RunnerOptions): RunnerFunction {
+  const { tsconfig, expectedAudits } = options;
+  return async (): Promise<AuditOutputs> => {
+    const diagnostics = await getTypeScriptDiagnostics({ tsconfig });
+    const result: Record<
+      CodeRangeName,
+      Pick<AuditReport, 'slug' | 'details'>
+    > = diagnostics.reduce(
+      (acc, diag) => {
+        const slug = tSCodeToAuditSlug(diag.code);
+        const existingIssues: Issue[] =
+          (acc[slug] && acc[slug].details?.issues) || ([] as Issue[]);
+        return {
+          ...acc,
+          [slug]: {
+            slug,
+            details: {
+              issues: [...existingIssues, getIssueFromDiagnostic(diag)],
+            },
+          },
+        };
+      },
+      {} as unknown as Record<
+        CodeRangeName,
+        Pick<AuditOutput, 'slug' | 'details'>
+      >,
+    );
```

Comment:
I see `as unknown as` brute-force casting :arrow_right: I complain :sweat_smile: 

The _true_ type of the `reduce` accumulator is actually a `Partial`, because not every code range need be emitted. Then the initial accumulator value is also valid (although a soft `as` would've been enough anyway).

There's also the question of where to put the accumulator type in `reduce`. In most situations, the type parameter is safest and most convenient (type used for implementation as well). There are some cases where the initial and final accumulator types are different, then it's a bit more tricky, but fortunately that's not the case here.

```suggestion
    const result = diagnostics.reduce<
      Partial<Record<CodeRangeName, Pick<AuditOutput, 'slug' | 'details'>>>
    >((acc, diag) => {
      const slug = tSCodeToAuditSlug(diag.code);
      const existingIssues: Issue[] =
        (acc[slug] && acc[slug].details?.issues) || ([] as Issue[]);
      return {
        ...acc,
        [slug]: {
          slug,
          details: {
            issues: [...existingIssues, getIssueFromDiagnostic(diag)],
          },
        },
      };
    }, {});
```